colexecjoin: improve hash joiner probing

yuzefovich opened the pull request:
This PR contains several commits that improve the hash joiner probing. The last commit contains an extensive comment about how different pieces used throughout the probing phase fit together. Overall, this PR improves the speed of the hash join by up to 30% with minor slowdowns in some cases in the microbenchmarks; there appear to be some minor changes for the hash aggregator and the unordered distinct, full colexec benchmarks are [here](https://gist.github.com/yuzefovich/0834cf0477abbd9b9ba0189970d15338).

Impact of the PR on TPCH queries (averaged over 100 runs of `tpchvec/perf`):
```
Q1:	before: 4.08s	after: 4.03s	 -1.14%
Q2:	before: 3.08s	after: 3.04s	 -1.27%
Q3:	before: 2.52s	after: 2.45s	 -2.80%
Q4:	before: 1.90s	after: 1.83s	 -3.59%
Q5:	before: 2.58s	after: 2.41s	 -6.30%
Q6:	before: 6.63s	after: 6.60s	 -0.54%
Q7:	before: 6.34s	after: 6.24s	 -1.56%
Q8:	before: 1.20s	after: 1.17s	 -2.85%
Q9:	before: 5.60s	after: 5.70s	 1.88%
Q10:	before: 2.27s	after: 2.15s	 -5.61%
Q11:	before: 0.69s	after: 0.68s	 -0.71%
Q12:	before: 7.13s	after: 7.07s	 -0.74%
Q13:	before: 1.26s	after: 1.26s	 -0.10%
Q14:	before: 0.68s	after: 0.66s	 -1.89%
Q15:	before: 3.83s	after: 3.82s	 -0.47%
Q16:	before: 0.93s	after: 0.91s	 -2.12%
Q17:	before: 0.27s	after: 0.27s	 -0.18%
Q18:	before: 2.45s	after: 2.48s	 1.11%
Q19:	before: 0.55s	after: 0.54s	 -1.15%
Q20:	before: 11.65s	after: 11.73s	 0.67%
Q21:	before: 7.22s	after: 7.13s	 -1.21%
Q22:	before: 0.61s	after: 0.59s	 -2.56%
```

Epic: None

yuzefovich committed:
colexecjoin: remove some duplication

This commit refactors `Check` method of the hash table to have the same
signature as `DistinctCheck` which allows us to remove some code
duplication in the hash joiner. Effectively, this is a no-op change.

Release note: None

yuzefovich committed:
colexecjoin: refactor hash joiner probing a bit

This commit changes the probing logic in the hash joiner. Previously, we
would always include each probing tuple into `ToCheck` slice, regardless
of the corresponding `ToCheckID` value. However, if that `keyID` is
zero, then it means that the corresponding probing tuple doesn't have
a hash match, and before this commit we would simply skip over such
tuple in the checker function. This commit makes it so that we ensure
that only probing tuples with hash matches are included into `ToCheck`
slice. In other words, we're lifting the non-zero `keyID` check from the
checker function into its caller which makes the code a bit easier to
follow. We recently made a similar change in the hash aggregator and the
unordered distinct, so this change unifies that approach.

This required a minor change so that the checker function doesn't update
`ToCheckID` to zero once it finds a NULL value in the probing tuple when
null equality is not allowed. Instead, we now mark this tuple accordingly
in the `differs` and `foundNull` slices. This also means that for
consequent equality columns we will now do the explicit comparison of
values for this probing tuple even though we know that it will be distinct.
We already have a TODO to address this, and it'll be done in the following
commit (that's why we marked it in `differs` in this commit). Marking it
in `foundNull` is needed so that we don't include this tuple for the next
Check iteration - the tuple won't ever get an equality match, so there is
no point in trying another "candidate".

[Benchmarks](https://gist.github.com/yuzefovich/37d53fc2647929901b9d01d3400ba161)
mostly show an improvement.

Additionally, this allows us to remove some special case behavior in the
hash joiner about populating the `ToCheck` slice initially.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/101954)
<!-- Reviewable:end -->

yuzefovich committed:
colexechash: short-circuit checking already different tuples

This commit modifies the equality check function of the hash table to
consult `differs` slice to see if the probing tuple is already different
from the "candidate" tuple in order to short-circuit the checking. This
allows us to avoid redundant evaluation of the comparisons between the
tuple elements. An additional improvement of this change is restoring
the ability to stop checking the probing tuple once a NULL value is
found in the equality column (which was lost in the previous commit).

[Benchmarks](https://gist.github.com/yuzefovich/b802ea46e966927aa9da4409a6238c03)
show expected but somewhat mixed results: namely, there is a slowdown
for single equality column case but speedup for three eq cols case;
also, the speedup is more pronounced for bytes (where the comparison is
more expensive). Overall, I believe this change is beneficial.

Release note: None

yuzefovich committed:
colexechash: add extensive comment about hash joiner probing

I've recently spent significant amount of time reading through the hash
joiner probing code, so this commit dumps all of the context into an
extensive comment with an example.

Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> I like the explanation, took me a while to parse it but I think I understand the hash table better now.

> Reviewed 3 of 3 files at r1, 6 of 6 files at r2, 2 of 4 files at r3, 2 of 2 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/101954#-:-NUhbkL0AaB-EvNlOgP9:b-el2b8x)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
